tests/end_to_end_tests/run_in_remote_machine/ec2.py

```diff
@@ -0,0 +1,361 @@
+        workflow_suffix = workflow_id or ""
+        name = f"{INSTANCE_NAME_STRING}-{workflow_suffix}-{ec2_image.image_name}"
+
+        logger.info("Starting node provisioning ...")
+
+        file_mappings[test_runner_path] = f"/tmp/test_runner"
+
+        file_deployment_steps = []
+        for source, dst in file_mappings.items():
+            file_deployment_steps.append(
+                FileDeployment(str(source), str(dst))
+            )
+
+        deployment = MultiStepDeployment(add=file_deployment_steps)
+
+        try:
+            node = driver.deploy_node(
+                name=name,
+                image=image,
+                size=size,
+                ssh_key=private_key_path,
+                ex_keyname=private_key_name,
+                ex_security_groups=[security_group],
+                ssh_username=ec2_image.ssh_username,
+                ssh_timeout=20,
+                max_tries=max_tries,
+                wait_period=15,
+                timeout=deploy_overall_timeout,
+                deploy=deployment,
+                at_exit_func=destroy_node_and_cleanup,
+            )
+        except DeploymentError as e:
+            print("Deployment failed: %s" % (str(e)))
+            node = e.node
+            stdout = getattr(e.original_error, "stdout", None)
+            stderr = getattr(e.original_error, "stderr", None)
+            raise Exception(f"Deployment is not successful.\nStdout: {stdout}\nStderr: {stderr}")
+
+        return node
+
+    def run_command():
+
+        ssh = paramiko.SSHClient()
+        ssh.load_system_host_keys()
+        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh.connect(
+            hostname=node.public_ips[0],
+            port=22,
+            username=ec2_image.ssh_username,
+            key_filename=str(private_key_path)
+        )
+
+        final_command = [
+            "/tmp/test_runner",
+            "-s",
+            *command
+        ]
+
+        command_str = shlex.join(final_command)
+        stdin, stdout, stderr = ssh.exec_command(
+            command=f"TEST_RUNS_REMOTELY=1 sudo -E {command_str} 2>&1",
+        )
+
+        print(f"stdout: {stdout.read().decode()}")
+
+        ssh.close()
+
+        assert stdout.channel.recv_exit_status() == 0, "Remote test execution has failed with."
+
+    file_mappings = file_mappings or {}
+    start_time = int(time.time())
+
+    driver_cls = get_driver(Provider.EC2)
+    driver = driver_cls(
+        access_key,
+        secret_key,
+        region=region
+    )
+
+    boto_client = boto3.client(
+        "ec2",
+        aws_access_key_id=access_key,
+        aws_secret_access_key=secret_key,
+        region_name=region
+    )
+
+    # Add current public IP to security group's prefix list.
+    # We have to update that prefix list each time because there are to many GitHub actions public IPs, and
+    # it is not possible to whitelist all of them in the AWS prefix list.
+    add_current_ip_to_prefix_list(
```

Comment:
This can be another TODO but it would be good to remove it when done (I'm assuming the security group will always have it & just end up collecting lots of ips otherwise)